File: 1381.design-a-stack-with-increment-operation.py
#
# @lc app=leetcode id=1381 lang=python3
#
# [1381] Design a Stack With Increment Operation
#

# @lc code=start
class CustomStack:

    def __init__(self, maxSize: int):
        self.stack = []
        self.count = 0
        self.maxSize = maxSize
    def push(self, x: int) -> None:
        if self.count < self.maxSize:
            self.stack.append(x)
            self.count += 1 
    def pop(self) -> int:
        if not self.stack:
            return -1
        self.count -= 1
        return self.stack.pop()
    def increment(self, k: int, val: int) -> None:
        if k > self.count:
            for i in range(self.count):
                self.stack[i] += val
        else:
            for i in range(k):
                self.stack[i] += val
        


# Your CustomStack object will be instantiated and called as such:
# obj = CustomStack(maxSize)
# obj.push(x)
# param_2 = obj.pop()
# obj.increment(k,val)
# @lc code=end

File: 1381.design-a-stack-with-increment-operation.java
/*
 * @lc app=leetcode id=1381 lang=java
 *
 * [1381] Design a Stack With Increment Operation
 */
// @lc code=start
class CustomStack {
    int[] stack;
    int[] inc;
    int curSize = 0;

    public CustomStack(int maxSize) {
        stack = new int[maxSize];
        inc = new int[maxSize];
    }

    public void push(int x) {
        if(curSize < stack.length){
            stack[curSize] = x;
            curSize++;
        }
    }

    public int pop() {
        if(curSize == 0) return -1;
        curSize--;
        int val = stack[curSize] + inc[curSize];
        if(curSize > 0) inc[curSize - 1] += inc[curSize];
        inc[curSize] = 0;
        return val;
    }

    public void increment(int k, int val) {
        int idx = Math.min(k, curSize) - 1;
        if(idx >= 0) inc[idx] += val;
    }
}

/**
 * Your CustomStack object will be instantiated and called as such:
 * CustomStack obj = new CustomStack(maxSize);
 * obj.push(x);
 * int param_2 = obj.pop();
 * obj.increment(k,val);
 */
// @lc code=end

